### PR TITLE
#63 freinet data import

### LINKED ISSUE #63: Fetch and parse information about accepting businesses from freinet
Belongs to #11

### PR DESCRIPTION
**Pull request by RoAnBu**
closes #63

### CODE REVIEW

**Comment by aiquita** [comment]
> 
> 
> closes #63

I think it is a first shot especially according to the technology. On the analysis side we have to have a deeper look into the data again. For testing purposes its suffices.

**Comment by RoAnBu** [comment]
> I think it is a first shot especially according to the technology. On the analysis side we have to have a deeper look into the data again. For testing purposes its suffices.
> 
> Moreover categories are not imported yet.

Yes, not finished yet. Therefore draft PR. I just like PRs for discussions and an overview of what's being worked on.

@aiquita Do you want to continue working on the branch? Otherwise i would continue

File: backend/build.gradle.kts [hunk 1]
```diff
@@ -49,9 +49,14 @@ dependencies {
     implementation("org.jetbrains.exposed", "exposed-core", exposed_version)
     implementation("org.jetbrains.exposed", "exposed-dao", exposed_version)
     implementation("org.jetbrains.exposed", "exposed-jdbc", exposed_version)
-    implementation("com.impossibl.pgjdbc-ng", "pgjdbc-ng", "0.8.6")
+    implementation("org.postgresql", "postgresql", "42.2.18")
```
Comment (Schedulaar) [inline backend/build.gradle.kts:52]: Why did you remove the impossibl?

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            .parse<FreinetData>(freinetJson)
+
+    private fun importAcceptingStores(acceptingStores: List<FreinetAcceptingStore>) {
+        println("Inserting data into db")
+        try {
+            transaction {
+                for (acceptingStore in acceptingStores) {
+                    val address = AddressEntity.new {
+                        street = acceptingStore.street
+                        houseNumber = "10"
```
Comment (Schedulaar) [inline backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt:49]: Is there no house number in the data? Maybe in the street name?

File: backend/src/main/resources/freinet_import/categories.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
+[
```
Comment (Schedulaar) [inline backend/src/main/resources/freinet_import/categories.json:1]: Don't we have an endpoint that we could fetch from instead of serving this statically? (Although it doesn't really matter)

**Review by Schedulaar — APPROVED** [review]
Minor:
You use the ```!!``` operator quite excessively... Afaik this should be avoided where possible. Maybe you can get rid of it at some places by using ```?: throw IllegalArgumentException``` or some more suitable exceptions

File: backend/build.gradle.kts (hunk 1 above)
Comment (RoAnBu) [reply backend/build.gradle.kts:52]: Has no implementation for postgis columns and they don't plan to implement it. The main postgres jdbc driver worked straight away.

Does the impossibl driver have any advantages we use/wanted to use? We could maybe use different drivers for data importing and querying but idk if that is a good idea

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt (hunk 2 above)
Comment (RoAnBu) [reply backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt:49]: Good question, didnt see that. @aiquita? 😄

File: backend/src/main/resources/freinet_import/categories.json (hunk 3 above)
Comment (RoAnBu) [reply backend/src/main/resources/freinet_import/categories.json:1]: No idea. @maxammann is there a way to see all files that can be queried from the freinet api?

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/database/Database.kt [hunk 4]
```diff
@@ -51,12 +51,10 @@ class Database {
...
 
     val db by lazy {
-        val ds = PGDataSource()
-        ds.serverName = System.getProperty("app.postgres.host")
-        ds.databaseName = System.getProperty("app.postgres.database")
-        ds.user = System.getProperty("app.postgres.user")
-        ds.password = System.getProperty("app.postgres.password")
-
-        connect(ds)
+        val connectionUrl = "jdbc:postgresql://" +
```
Comment (maxammann) [inline backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/database/Database.kt:54]: what does reWriteBatchedInserts do? Do we require it?

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                        postalCode = acceptingStore.postalCode
+                        locaction = acceptingStore.location
+                        state = "de"
```
Comment (maxammann) [inline backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/AcceptingStoresImporter.kt:52]: Is state fine here or should it be "Germany"? We should define what should be in this field.

E.g. Alpha-2 codes: https://www.iban.com/country-codes

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/annotations/BooleanField.kt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package xyz.elitese.ehrenamtskarte.importer.annotations
+
+@Target(AnnotationTarget.FIELD)
+annotation class BooleanField
```
Comment (maxammann) [inline backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/annotations/BooleanField.kt:4]: Minor: Kind of weird that these are required. Shouldn't Klaxon support this?

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/dataqa/FreinetDataFilter.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package xyz.elitese.ehrenamtskarte.importer.dataqa
+
+import xyz.elitese.ehrenamtskarte.importer.types.FreinetAcceptingStore
+
+object FreinetDataFilter {
+    fun filterOutInvalidEntries(freinetAcceptingStores : List<FreinetAcceptingStore>) : List<FreinetAcceptingStore> {
```
Comment (maxammann) [inline backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/dataqa/FreinetDataFilter.kt:6]: Are we sure that these should be filtered out? Can you post a list of partners which have invalid data?

Maybe we should transform the data to be valid instead of filtering the entries out.

File: backend/src/main/resources/freinet_import/categories.json (hunk 3 above)
Comment (maxammann) [reply backend/src/main/resources/freinet_import/categories.json:1]: Not that I know of. This seems fine!

Comment (maxammann) [reply backend/src/main/resources/freinet_import/categories.json:1]: But thee ids seems wrong. They go from 0 to 8 in NRW.

**Review by maxammann — APPROVED** [review]

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/database/Database.kt (hunk 4 above)
Comment (RoAnBu) [reply backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/database/Database.kt:54]: It tries to batch following inserts together to make the inserts faster. However that does not seem to work because we have inserts that depend on each other. It was a try, I can remove it again I guess

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/annotations/BooleanField.kt (hunk 6 above)
Comment (RoAnBu) [reply backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/annotations/BooleanField.kt:4]: I think the issue is that all fields in the original json are strings, so they need to be converted to bools, ints etc. @aiquita can you confirm that this was the reason?

File: backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/dataqa/FreinetDataFilter.kt (hunk 7 above)
Comment (RoAnBu) [reply backend/src/main/kotlin/xyz/elitese/ehrenamtskarte/importer/dataqa/FreinetDataFilter.kt:6]: Currently i think we have less than 10 invalid entries, so not that urgent. Most of them should be fixable, however some are missing the category id which could be a problem. I think we need a concept for broken data later anyway and we also need to search for duplicates in the database etc. I think a separate user story/technical task would be good

File: backend/src/main/resources/freinet_import/categories.json (hunk 3 above)
Comment (RoAnBu) [reply backend/src/main/resources/freinet_import/categories.json:1]: Freinet api returns values from 1-9 (inclusive), I got the values from here: https://github.com/ehrenamtskarte/artefacts/blob/main/api_access.md

**Review by aiquita — APPROVED** [review]
